File: elixir/507-PerfectNumber/main.exs
defmodule Solution do
  @spec check_perfect_number(num :: integer) :: boolean
  def check_perfect_number(num) do
    upper = floor(:math.sqrt(num))

    nums = for x <- 2..upper, rem(num, x) == 0, do: [x, div(num, x)]
    total = [[1] | nums]
            |> Enum.flat_map(fn x -> x end)
            |> Enum.reduce(fn n, acc -> n + acc end)
    total == num
  end
end

IO.inspect(Solution.check_perfect_number(28))
IO.inspect(Solution.check_perfect_number(7))

File: cpp/507-PerfectNumber/main.cpp
#include <cmath>

using namespace std;

class Solution 
{
public:
    bool checkPerfectNumber(int num) 
    {
        if(num < 2) return false;

        auto upper = static_cast<int>(sqrt(num));
        auto sum = 1;

        for(auto i = 2; i <= upper; ++i)
        {
            if(num % i == 0)
            {
                sum += i;
                sum += num / i;
            }
        }

        return num == sum;
    }
};

int main()
{
    Solution s;
    auto test1 = s.checkPerfectNumber(28);
    auto test2 = s.checkPerfectNumber(6);
}